internal/authentication/oidc/clientcredentials.go

```diff
@@ -0,0 +1,213 @@
+/*
+Copyright 2021 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package oidc
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"golang.org/x/oauth2"
+	ccreds "golang.org/x/oauth2/clientcredentials"
+	"k8s.io/utils/clock"
+
+	"github.com/dapr/kit/logger"
+)
+
+const (
+	oidcScopeOpenID = "openid"
+)
+
+type ClientCredentialsOptions struct {
+	Logger       logger.Logger
+	TokenURL     string
+	ClientID     string
+	ClientSecret string
+	Scopes       []string
+	Audiences    []string
+	CAPEM        []byte
+}
+
+// ClientCredentials is an OAuth2 Token Source that uses the client_credentials
+// grant type to fetch a token.
+type ClientCredentials struct {
+	log          logger.Logger
+	currentToken *oauth2.Token
+	httpClient   *http.Client
+	fetchTokenFn func(context.Context) (*oauth2.Token, error)
+
+	lock    sync.RWMutex
+	wg      sync.WaitGroup
+	closeCh chan struct{}
+	closed  atomic.Bool
+	clock   clock.Clock
+}
+
+func NewClientCredentials(ctx context.Context, opts ClientCredentialsOptions) (*ClientCredentials, error) {
+	conf, httpClient, err := toConfig(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	token, err := conf.Token(context.WithValue(ctx, oauth2.HTTPClient, httpClient))
+	if err != nil {
+		return nil, fmt.Errorf("error fetching initial oidc client_credentials token: %w", err)
+	}
+
+	opts.Logger.Info("Fetched initial oidc client_credentials token")
+
+	return &ClientCredentials{
+		log:          opts.Logger,
+		currentToken: token,
+		httpClient:   httpClient,
+		closeCh:      make(chan struct{}),
+		clock:        clock.RealClock{},
+		fetchTokenFn: conf.Token,
+	}, nil
+}
+
+func (c *ClientCredentials) Run(ctx context.Context) {
+	c.log.Info("Running oidc client_credentials token renewer")
+	renewDuration := c.tokenRenewDuration()
+
+	c.wg.Add(1)
+	go func() {
+		defer func() {
+			c.log.Info("Stopped oidc client_credentials token renewer")
+			c.wg.Done()
+		}()
+
+		for {
+			select {
+			case <-c.closeCh:
+				return
+			case <-ctx.Done():
+				return
+			case <-c.clock.After(renewDuration):
```

Comment:
This is anti-pattern for using OAuth and we should not refresh tokens in background. Instead, tokens should be fetched lazily and refreshed only if necessary. This is especially true when the consumer of the tokens is something like Pulsar where connections are long-lived so tokens don't need to be refreshed constantly.

Reply:
Why is this an anti-pattern? I have never heard of this. Fetching tokens lazily is strictly worse from the dapr perspective, increasing latency for the pubsub app which is particularly acute for publishing. I don't think we should be concerned with reducing load on the OIDC provider- signing JWTs is not resource intensive and is basically free for all intense and purposes. Regardless of long lived etc. we can assume that lazily requesting tokens will produce about roughly the same amount of requests as renewing in the background for a publishing app (bar the renewal time we agree on), however fetching tokens lazily will increase the latency on the unlucky publish(es).

Reply:
> So refreshing automatically "just in case" seems really wasteful.

Fair enough- I still disagree with this statement though. I maintain that the latency improvements for daprd fair out way the small resource used to fetch more tokens than perhaps needed.

> It may help hitting rate-limits. It also can create additional noise in the audit logs that are sent to SIEMs, etc.

An IdP which rate limits clients for requesting tokens at half expiry is miss configured imo, and audit logs are built around the concept of clients renewing their identity documents. They will be indexed and filtered around this exact scenario.